scipy/stats/mstats_basic.py

```diff
@@ -543,21 +543,40 @@ def _spearmanr_2cols(x):
         return SpearmanrResult(rs, prob)
 
 
-def _kendall_p_exact(n, c):
-    # Exact p-value, see Maurice G. Kendall, "Rank Correlation Methods" (4th Edition), Charles Griffin & Co., 1970.
+def _kendall_p_exact(n, c, alternative='two-sided'):
+
+    # Use the fact that distribution is symmetric: always calculate a CDF in
+    # the left tail.
+    # This will be the one-sided p-value if `c` is on the side of
+    # the null distribution predicted by the alternative hypothesis.
+    # The two-sided p-value will be twice this value.
+    # If `c` is on the other side of the null distribution, we'll need to
+    # take the complement and add back the probability mass at `c`.
+    in_right_tail = (c >= (n*(n-1))//2 - c)
+    alternative_greater = (alternative == 'greater')
+    c = int(min(c, (n*(n-1))//2 - c))
+
+    # Exact p-value, see Maurice G. Kendall, "Rank Correlation Methods"
+    # (4th Edition), Charles Griffin & Co., 1970.
     if n <= 0:
         raise ValueError(f'n ({n}) must be positive')
     elif c < 0 or 4*c > n*(n-1):
         raise ValueError(f'c ({c}) must satisfy 0 <= 4c <= n(n-1) = {n*(n-1)}.')
     elif n == 1:
         prob = 1.0
+        p_mass_at_c = 1
     elif n == 2:
         prob = 1.0
+        p_mass_at_c = 0.5
     elif c == 0:
         prob = 2.0/math.factorial(n) if n < 171 else 0.0
+        p_mass_at_c = prob/2
     elif c == 1:
         prob = 2.0/math.factorial(n-1) if n < 172 else 0.0
-    elif 4*c == n*(n-1):
+        p_mass_at_c = (n-1)/math.factorial(n)
```

Comment:
There's are `factorial(n)` permutations. There are `n-1` for which `c=1` (swap adjacent ranks), so the probability of `c=1` under the null hypothesis is `(n-1)/factorial(n)`.
Another way to look at it is that the two-sided p-value is `2.0/math.factorial(n-1)` = `2*n/math.factorial(n)`. Half of that is the probability of `c=0` OR `c=1` under the null hypothesis, and `1/factorial(n)` of that is associated with `c=0`, so `(n-1)/factorial(n)` is associated with `c=1`.